refactor!: use LibraryCollectionLocator and LibraryContainerLocator [FC-0083]

pomegranited opened the pull request:
<!--

Note: Please refer to the Support Development Guidelines on the wiki page to consider backporting to active releases:
https://openedx.atlassian.net/wiki/spaces/COMM/pages/4248436737/Support+Guidelines+for+active+releases

Please give your pull request a short but descriptive title.
Use conventional commits to separate and summarize commits logically:
https://open-edx-proposals.readthedocs.io/en/latest/oep-0051-bp-conventional-commits.html

Use this template as a guide. Omit sections that don't apply.
You may link to information rather than copy it, but only if the link is publicly readable.
If the linked information must be private (because it contains secrets), clearly label the link as private.

-->

## Description

Refactors the content_libraries.api to use LibraryCollectionLocator and LibraryContainerLocator keys, instead of passing separate LibraryUsageKeyV2 keys along with the collection/container locators.

Renames misleading uses of `collection_usage_key` to `collection_locator`, including in the `content_libraries.api` and `content.search.api` method names and parameters.

This change impacts Developers, but should not affect end users.

This refactoring seems reasonable to do without going through deprecation, given the minimal use of these APIs.

## Supporting information

cf https://github.com/openedx/openedx-events/pull/479#discussion_r1998824771
Blocked by: https://github.com/openedx/openedx-events/pull/487 (merged)
Private-ref: [FAL-4120](https://tasks.opencraft.com/browse/FAL-4120)

## Testing instructions

Code review + CI should be sufficient for verifying this change.

Manual testing with the Authoring MFE:

* Ensure you can create, update, and delete library collections, and that these changes show up as expected in the MFE.
* Repeat ^ for containers (update and delete may not be available yet)

## Deadline

ASAP -- want to get this in before Teak.

pomegranited committed:
refactor: split collections and courseware_import code out of api/libraries.py

pomegranited committed:
refactor: moves library blocks code into blocks modules

pomegranited committed:
fix: trigger LIBRARY_CONTAINER_UPDATED on restore_library_block

pomegranited committed:
fix: container tests not sending signals

pomegranited committed:
refactor: move api exceptions into own file

pomegranited committed:
Merge remote-tracking branch 'origin/master' into jill/libraries-api-refactor

pomegranited committed:
temp: depends on openedx-events WIP branch

see https://github.com/openedx/openedx-events/pull/487

pomegranited committed:
refactor!: use LibraryCollectionLocator

instead of the LibraryUsageKeyV2 + collection.key in the following
functions:

* openedx_events LibraryCollectionData
* update_library_collection_index_doc
* upsert_library_collection_index_doc
* searchable_doc_for_collection
* searchable_doc_tags_for_collection
* upsert_collection_tags_index_docs
* update_library_components_collections

Also renames uses of collection_usage_key to collection_locator in
content_libraries.api:

* get_library_collection_usage_key -> library_collection_locator
* get_library_collection_from_usage_key -> get_library_collection_from_locator

openedx-webhooks commented:
Thanks for the pull request, @pomegranited!


This repository is currently maintained by `@openedx/wg-maintenance-edx-platform`.


Once you've gone through the following steps feel free to tag them in a comment and let them know that your changes are ready for engineering review.

<details><summary>:radio_button: Get product approval</summary>

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.
</details>

<details><summary>:radio_button: Provide context</summary>

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads
</details>





<details><summary>:radio_button: Get a green build</summary>

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.
</details><details>


---

<details><summary>Where can I find more information?</summary>

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)
</details>

<details><summary>When can I expect my changes to be merged?</summary>

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*
</details>
<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

pomegranited committed:
refactor!: use LibraryContainerLocator

instead of the LibraryUsageKeyV2 + container_key in LibraryContainerData
and content.search.handlers.update_library_container_index_doc

rpenido reviewed (approved):
LGTM :+1:
Thank you for your work, @pomegranited!

- [x] I tested this using the CI pipeline
- [x] I read through the code
- [x] I checked for accessibility issues
- [x] Includes documentation

Let me know if you want another check after merging and fixing the conflicts!

pomegranited committed:
Merge remote-tracking branch 'origin/master' into jill/container-collection-keys

pomegranited committed:
fix: pass strings to tasks, not keys

ChrisChV commented:
@pomegranited When creating a Unit, I got this error:

```
cms-1               | Traceback (most recent call last):
cms-1               |   File "/openedx/venv/lib/python3.11/site-packages/celery/app/trace.py", line 453, in trace_task
cms-1               |     R = retval = fun(*args, **kwargs)
cms-1               |                  ^^^^^^^^^^^^^^^^^^^^
cms-1               |   File "/openedx/venv/lib/python3.11/site-packages/celery/app/autoretry.py", line 38, in run
cms-1               |     return task._orig_run(*args, **kwargs)
cms-1               |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
cms-1               |   File "/openedx/venv/lib/python3.11/site-packages/edx_django_utils/monitoring/internal/code_owner/utils.py", line 195, in new_function
cms-1               |     return wrapped_function(*args, **kwargs)
cms-1               |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
cms-1               |   File "/openedx/edx-platform/openedx/core/djangoapps/content/search/tasks.py", line 128, in update_library_container_index_doc
cms-1               |     library_key = container_key.library_key
cms-1               |                   ^^^^^^^^^^^^^^^^^^^^^^^^^
cms-1               | AttributeError: 'str' object has no attribute 'library_key'
```

pomegranited committed:
fix: update_library_container_index_doc task takes a string

pomegranited committed:
Merge remote-tracking branch 'origin/master' into jill/container-collection-keys

pomegranited commented:
> When creating a Unit, I got this error:

Thank you for catching this @ChrisChV ! My last-minute 71afe3f6c5ff5a78351b828b9d1dd15dc005ff93 caused this problem, so have fixed with c4329eed4d25b68bacafeb542c39b0b5cb1a326c. Can you please re-test?

pomegranited commented on requirements/edx/kernel.in:
```diff
@@ -115,7 +115,7 @@ olxcleaner
 openedx-calc                        # Library supporting mathematical calculations for Open edX
 openedx-django-require
-openedx-events                      # Open edX Events from Hooks Extension Framework (OEP-50)
+git+https://github.com/openedx/openedx-events.git@refs/pull/487/head
```
TODO: update requirement when https://github.com/openedx/openedx-events/pull/487 merged.

pomegranited committed:
feat: updates openedx-events dependency
